python/ray/experimental/sgd/sgd.md

```diff
@@ -0,0 +1,401 @@
+    ...
+    return model
+
+trainer = Trainer(num_workers=8, backend="torch")
+dataset = ray.dataset.filter().pipeline(length=50) # do not call split!
+trainer.run(train_func, config={}, dataset=dataset)
+
+```
+
+Cross Validation:
+
+```python
+
+cv_splits = ray.dataset.to_cv(splits=5)
+
+for train_set, val_set in cv_splits:
+    result = trainer.run(train_func, dataset=train_set)
+    result = trainer.run(train_func, dataset=val_set)
+```
+
+<details>
+<summary> Alternative API </summary>
+
+```python
+# This was not chosen because it requires the user to think about 
+# too many points - both the training function signature and the Trainer declaration.
+
+# def train_func(*args, dataset=None, **kwargs,):
+#     if ray.sgd.in_worker():
+#         dataloader = dataset.get_shard(self.rank).to_torch()
+#     else:
+#         dataset = ImageDataset("code")
+#         dataloader = DataLoader(dataset)
+#     for i, (x, y) in dataloader:
+#         do_x_y(x, y)
+
+# Alternative API: SPMD call.
+# This was not chosen because it doesn't allow users to dynamically shard the dataset.
+# # Trainer will make a SPMD call.
+# model = trainer.map(
+#     training_func, 
+#     data=splits, 
+#     arg1=[X for _ in trainer.workers], 
+#     arg2=[Y for _ in trainer.workers])
+# trainer.run(validation_func, model=model, dataset=val_dataset)
+
+
+# Option 3:
+# splits = ray.dataset.filter().split(locality_hint=trainer.workers)
+# trainer.run(train_func, dataset=dataset)
+```
+</details>
+
+
+Intermediate Reporting
+----------------------
+
+Satisfies: R3 (partially), R7.
+
+Provide support for callbacks (which can be tensorboard, wandb etc). Also provides the same interface to use with Tune.
+
+```python
+def train_func(config):
+    ...
+    for _ in config["epochs"]:
+        metrics = train()
+        metrics = validate(...)
+        ray.sgd.report(**metrics)
+    return model
+
+# Higher level API
+results = trainer.run(train_func, config=config, callbacks=[...])
+
+# Lower-level generator API
+generator = trainer.run_generator(train_func, config=config)
+
+for result in generator:
+    do_stuff(result)
+
+assert generator.is_finished()
+model = generator.get_returns()
+
+
+```
+
+
+
+Hyperparameter Tuning
+---------------------
+
+Satisfies: R3.
+
+Specific requirements for hyperparameter tuning.
+
+1. Users should be able to do distributed tuning, training, and data processing.
+2. Users should be able to tune the parameters of the dataset.
+3. Going from SGD -> Tune should be easy.
+4. Ability to do cross validation (1 dataset, shard multiple ways)
+
+In the easy case, users can Tune it with a 2 line code change.
+
+```python
+def training_func(config):
+    if ray.sgd.in_worker():
+        dataloader = ray.sgd.get_dataset()\
+            .get_shard(torch.rank())\
+            .to_torch(batch_size=config["batch_size"])
+    else:
+        ...
+
+    for i in epochs:
+        ray.sgd.report(...)  # use same intermediate reporting API
```

Comment:
Why is this ray.sgd instead of ray tune?

Reply:
Ah so the reason for this is that Tune will have multiple trials, with each trial having multiple SGD workers.

So `sgd.report()` is run in each SGD worker and results are reported to the SGD driver (which is a single Tune trial). The SGD driver can then do some aggregation of the results from each worker (for example, averaging) and then will call `tune.report`. This latter part will all happen internally.